### DISCUSSION TITLE
Dependency lookup blocked by IP safelisting for `bridgecrewio/checkov-action`

### DISCUSSION THREAD

**Discussion by patheard** [renovatebot/renovate#17826] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

We're getting a dependency lookup failure for the [`bridgecrewio/checkov-action`](https://github.com/bridgecrewio/checkov-action) GitHub Action that used to work.  From what we can see, IP safelisting is causing the problem and blocking the Renovate lookup:

```javascript
"response": {
    "statusCode": 403,
    "statusMessage": "Forbidden",
    "body": {
        "message": "Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and 18.237.254.32 is not permitted to access this resource.",
    }
    // etc.
}
```

The full log output can be seen in this job log: 
* `github/cds-snc/scan-files/814142718`

I'm assuming all we can do is add this dependency to our `ignoreDeps` config if we don't want to see the warning on PRs and in [our dependency dashboard](https://github.com/cds-snc/scan-files/issues/52).

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Unknown GitHub error
{
  "err": {
    "name": "HTTPError",
    "code": "ERR_NON_2XX_3XX_RESPONSE",
    "timings": {
      "start": 1663274118042,
      "socket": 1663274118042,
      "lookup": 1663274118242,
      "connect": 1663274118242,
      "secureConnect": 1663274118260,
      "upload": 1663274118260,
      "response": 1663274118425,
      "end": 1663274118434,
      "phases": {
        "wait": 0,
        "dns": 200,
        "tcp": 0,
        "tls": 18,
        "request": 0,
        "firstByte": 165,
        "download": 9,
        "total": 392
      }
    },
    "message": "Response code 403 (Forbidden)",
    "stack": "HTTPError: Response code 403 (Forbidden)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)",
    "options": {
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "accept": "application/vnd.github.machine-man-preview+json",
        "authorization": "***********",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://api.github.com/repos/bridgecrewio/checkov-action/tags?per_page=100",
      "hostType": "github-tags",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    },
    "response": {
      "statusCode": 403,
      "statusMessage": "Forbidden",
      "body": {
        "message": "Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and 18.237.254.32 is not permitted to access this resource.",
        "documentation_url": "https://docs.github.com/rest/reference/repos#list-repository-tags"
      },
      "headers": {
        "server": "GitHub.com",
        "date": "Thu, 15 Sep 2022 20:35:18 GMT",
        "content-type": "application/json; charset=utf-8",
        "transfer-encoding": "chunked",
        "x-github-media-type": "github.v3; param=machine-man-preview; format=json",
        "x-ratelimit-limit": "15000",
        "x-ratelimit-remaining": "14907",
        "x-ratelimit-reset": "1663277402",
        "x-ratelimit-used": "93",
        "x-ratelimit-resource": "core",
        "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
        "access-control-allow-origin": "*",
        "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
        "x-frame-options": "deny",
        "x-content-type-options": "nosniff",
        "x-xss-protection": "0",
        "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
        "content-security-policy": "default-src 'none'",
        "vary": "Accept-Encoding, Accept, X-Requested-With",
        "content-encoding": "gzip",
        "x-github-request-id": "A34A:775E:6A6414:6D548F:63238C86",
        "connection": "close"
      },
      "httpVersion": "1.1"
    }
  }
}
DEBUG: Datasource unauthorized
{
  "datasource": "github-tags",
  "packageName": "bridgecrewio/checkov-action",
  "url": "https://api.github.com/repos/bridgecrewio/checkov-action/tags?per_page=100"
}
DEBUG: Failed to look up dependency bridgecrewio/checkov-action (bridgecrewio/checkov-action)(packageFile=".github/workflows/terraform-security-scan.yml", dependency="bridgecrewio/checkov-action")
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by viceice** [comment]
This isn't a renovate bug. It's a github restriction and renovate can't skip this.

Also renovate app worker don't have fixed IP's (they will wandomly changed on worker updates), so they can't be whitelisted in any way.

**Comment by rarkins** [comment]
Bridgecrew are going to need to remove this restriction or separate their open source/ closed source into different orgs

**Reply by patheard** [reply]
That's what I figured - thanks for confirming.  I'll reach out to Bridgecrew.

**Reply by gruebel** [reply]
I don't see a problem here. We have the IP whitelisting for write permissions to the repo for several months in place. So if it worked a few days/weeks ago, then either something was adjusted in renovate or GitHub itself. The repo is public for any read interactions, so I don't understand the need to change anything?

**Reply by patheard** [reply]
Reading [Managing allowed IP addresses for a GitHub App](https://docs.github.com/en/enterprise-cloud@latest/developers/apps/building-github-apps/managing-allowed-ip-addresses-for-a-github-app), it's likely this:

> When an organization has an allow list, third-party applications that connect via a GitHub App will be denied access unless both of the following are true: 
<br>1. The creator of the GitHub App has configured an allow list for the application that specifies the IP addresses at which their application runs. See below for details of how to do this.
<br>2. The organization owner has chosen to permit the addresses in the GitHub App's allow list to be added to their own allow list. For more information, see "[Managing allowed IP addresses for your organization](https://docs.github.com/en/enterprise-cloud@latest/organizations/keeping-your-organization-secure/managing-allowed-ip-addresses-for-your-organization#using-github-actions-with-an-ip-allow-list)."

Based on [the previous response](https://github.com/renovatebot/renovate/discussions/17826#discussioncomment-3659471) that the Renovate app worker IP addresses are dynamic and cannot be safe-listed, I would take this to mean that the Renovate GitHub App won't work with orgs that have IP allow lists.

And it may very well have been months since this worked for us - I don't have an exact date when I started seeing the warnings in our Renovate logs.

**Reply by rarkins** [reply]
Renovate does not perform any write requests to third party repos, and would be denied for other permissions reasons if it ever tried. It's sounding like GitHub might have a bug when it comes to GitHub Apps + IP whitelisting, but @patheard we'd need a public reproduction repo demonstrating this problem before I could debug and be more confident of that. Seems like GitHub is denying read requests, while according to @gruebel it should only be applying to write requests.

**Reply by patheard** [reply]
Makes sense for the public reproduction repo before being able to debug.  

The way I read the docs, GitHub is not distinguishing between app read/write requests and only references the existence of an organization's IP allow list triggering the 2 access condition checks.

**Reply by patheard** [reply]
Had a chance to create some test cases, and there's definitely something odd going on with how the GitHub API treats the dependency lookup based on the source repo.  

Here's the test scenario:

1. Created [renovate-ip-allow-list](https://github.com/renovate-ip-allow-list) org with an IP allow list open to everyone `0.0.0.0/0` and `::0`.
2. Created [renovate-ip-allow-list/hello-action](https://github.com/renovate-ip-allow-list/hello-action) GitHub action in the org.
3. Onboarded Renovate to personal and org repos:
   1. ✅  [Personal repo](https://github.com/patheard/test-renovate-with-ip-allow-list): dependency lookup [succeeds](https://app.renovatebot.com/dashboard#github/patheard/test-renovate-with-ip-allow-list/817571029).
   2. ❌ [Organization repo](https://github.com/dto-btn/ansible-dev-laptop): dependency lookup fails with [unauthorized message referencing IP allow list](https://app.renovatebot.com/dashboard#github/dto-btn/ansible-dev-laptop/817570849).

I'll reach out to GitHub support and leave my `renovate-ip-allow-list` org around until the Enterprise Cloud trial runs out.

**Reply by patheard** [reply]
GitHub support got back to me and this is expected behaviour:

>  What's happening is that GitHub Apps run in their own security context, not as a user. Additionally, for public repositories IP allow lists only apply to organization members. 

☝🏻 which explains why look ups from my personal repo work, but fail from my cds-snc org repos.

> With the personal repo, the App installation is not an organization member since it isn't installed on the organization, but on a personal repository. While this is allowed, this App install should only be able to access resources which are already public, since it's running in its own non-member security context.

They also confirmed that for an org repo lookup on another org repo with an IP allow list, the only fix is the one described in [the docs](https://docs.github.com/en/enterprise-cloud@latest/developers/apps/building-github-apps/managing-allowed-ip-addresses-for-a-github-app).

**Reply by rarkins** [reply]
> this App install should only be able to access resources which are already public, since it's running in its own non-member security context.

Sure, but isn't the app only trying to access public resources? e.g. the tags or releases list

**Reply by patheard** [reply]
Yup, totally agreed - I don't know why GitHub is treating the API requests differently for personal repos vs org repos when it comes to IP allow lists.

**Reply by rarkins** [reply]
I really wonder if they don't understand the question we're asking. It definitely seems like a bug to me. For example, when bridgecrew set up IP allow lists I'm pretty sure they were hoping to block apps from fetching a tags list from public repositories unless the app uses fixed IP addresses.

**Reply by patheard** [reply]
At first they did misunderstand my question, but I was pretty clear in outlining the test case and failures/success we're seeing. Here's my thread with support:

----

Allie commented [2 days ago](https://support.github.com/ticket/personal/0/xxx#tc-9468870482324)
Hi Pat,

Thanks so much for clarifying this - this is what I suspected as the alternative case if I had misunderstood the initial case. What's happening is that GitHub Apps run in their own security context, not as a user. Additionally, for public repositories IP allow lists only apply to organization members.

In case four, the App installation is not an organization member since it isn't installed on the organization, but on a personal repository. The fact that it's on a personal repository of an organization member doesn't change this.

While this is allowed, this App install should only be able to access resources which are already public, since it's running in its own non-member security context.

Please let me know if you have any questions.

Cheers,

Allie

Pat Heard
Pat Heard commented [2 days ago](https://support.github.com/ticket/personal/0/xxx#tc-9468478684692)
Totally understandable - this is a weird case. The scenario we're seeing is:

1. [bridgecrewio](https://github.com/bridgecrewio) org has an IP allow list setup.
1. my [cds-snc](https://github.com/cds-snc) org uses the [Renovate GitHub app](https://github.com/apps/renovate) (3rd party) to lookup dependency updates.
1. Renovate is blocked in my cds-snc repos from looking up [bridgecrewio/checkov-action](https://github.com/bridgecrewio/checkov-action) tags which make sense since the 3rd party app is being blocked by the IP allow list (matches what the docs describe).
1. If I install Renovate in a [personal repo](https://github.com/patheard/test-renovate-with-ip-allow-list), it is not blocked from looking up [bridgecrewio/checkov-action](https://github.com/bridgecrewio/checkov-action) tags.

I would have expected case 4 to be blocked as well since it is still a 3rd party app attempting to query an org repo that has an IP allow list defined.

Allie
GitHub Support
Allie commented [2 days ago](https://support.github.com/ticket/personal/0/xxx#tc-9468322927252)
Hi Pat,

I may have misunderstood the report. The IP allow list is based on the repository being accessed, not on the App. From the report, my understanding is that the App can access a personal repository fine (this is to be expected, as IP allow lists don't apply to personal repositories), but can't access an organization repository.

Is this correct? If not, any additional clarification of exactly which App installs can and can't access which repositories would be helpful. I suspect it may be related to IP allow lists applying to organization members but not to non-members (since an App installed outside of an organization wouldn't be considered an organization member).

Kind regards,

Allie

Pat Heard
Pat Heard commented [2 days ago](https://support.github.com/ticket/personal/0/xxx#tc-9468006218900)
Thanks for getting back to me Allie. Just to make sure I understand the difference in 3rd party app behaviour is this correct?

Personal repo: 3rd party app will successfully query an org repo with an IP allow list.
Org repo: 3rd party app will fail to query an org repo with an IP allow list.
Really what I'm curious about is why 3rd party apps are behaving differently if they're install in a personal repo vs an org repo.

Allie
GitHub Support
Allie commented [2 days ago](https://support.github.com/ticket/personal/0/xxx#tc-9466723918100)
Hi Pat,

Personal repositories wouldn't be covered by an organization's IP allow list. There's no way to bypass the IP allow list for a specific user account, so you'd need to add these IP addresses if you wish to continue using IP allow lists on this organization.

Please let me know if you have any questions, or if there's anything else I can help with!

Kind regards,

Allie

Pat Heard
Pat Heard commented [3 days ago](https://support.github.com/ticket/personal/0/xxx#tc-9449617965716)
Hi,

I've noticed odd behaviour that I'm hoping you can help explain. The issue is related to the [Renovate GitHub App](https://github.com/apps/renovate) being blocked from performing dependency lookups:

https://github.com/renovatebot/renovate/discussions/17826

When Renovate is running from an org repo and attempts to lookup a repo that is part of another org with an IP allow list, the lookup is blocked with a 403 response. [Reading the docs](https://docs.github.com/en/enterprise-cloud@latest/developers/apps/building-github-apps/managing-allowed-ip-addresses-for-a-github-app), it looks like Renovate would need to add all their app's IP addresses, and the org with the IP allow list would need to permit those IP addresses.

The odd part is that if Renovate is running in a personal repo, the same lookup succeeds and is not blocked by the IP allow list. Here's the test case I setup showing the behaviour:

https://github.com/renovatebot/renovate/discussions/17826#discussioncomment-3674979

Any idea why the lookup fails for the org repo but not the personal repo?

Also, is there anyway around the IP allow list blocking the read-only lookups from the third party app?

Thanks!
-pat

**Reply by rarkins** [reply]
Can you reproduce this in a public repo?

**Reply by patheard** [reply]
Do you mean a public personal repo?  I created this test case, which has the Renovate dependency lookup on `bridgecrewio/checkov-action` working fine:
https://github.com/patheard/test-renovate-with-ip-allow-list

On this public organization repo, the same lookup is failing:
https://github.com/dto-btn/ansible-dev-laptop

**Reply by rarkins** [reply]
Very interesting! I forked to https://github.com/renovate-reproductions/17826/issues/2 and also get the lookup error. Same repo forked to my personal space works: https://github.com/rarkins/ansible-dev-laptop/issues/3

Surely this is a mistake. IP lists are for security reasons. There's no security if some data is available to personal accounts but not orgs.

**Reply by patheard** [reply]
I agree - I don't see a valid reason for the IP allow list to block the lookup of public data.

**Reply by rarkins** [reply]
Not sure what to say except I'm pretty sure GitHub have a bug and probably aren't aware of it, despite your attempts

**Reply by morremeyer** [reply]
We encountered this issue after switching to renovate on-prem. Since we have a contract with GitHub, I'll reach out to them to make them aware and find a solution. Will keep everyone updated here.

**Reply by morremeyer** [reply]
Quick update: I'm still in contact with GitHub support to find the root cause or ideally resolve this. It's taking a bit since this involves different teams on GitHub's side depending on which part of this we're poking at/how we're looking at it. Still on it 😊

**Reply by morremeyer** [reply]
**Unauthenticated** requests for the public resources (in this case tags) for the repository are definitely not affected by this. Is there any way we could **disable** auth completely for requests to this repository? 

I checked the `hostRules` documentation and don't think there is a way to do this right now. Any other ideas?

**Reply by rarkins** [reply]
Renovate uses GraphQL to fetch tags and release so it must be authenticated to succeed

**Reply by morremeyer** [reply]
Quick reproduction script in Python.

You need `jwt` and `requests` installed in e.g. a virtualenv for this to work.

Change the config in the `Set config in the following block` section with the values for your setup

<details><summary>Script</summary>

```py
#!/usr/bin/env python3

import jwt
import requests
import time
import json

class graphQL:

    # Authorization is set in __init__
	headers = {}

	def __init__(self, token):
		super(graphQL, self).__init__()
		self.headers["Authorization"] = f"token {token}"

	def run_query(self, query, variables): # A simple function to use requests.post to make the API call. Note the json= section.
		request = requests.post('https://api.github.com/graphql', json={'query': query, 'variables': variables}, headers=self.headers)
		if request.status_code == 200:
			return request.json()
		else:
			raise Exception("Query failed to run by returning code of {}. Message {}".format(request.status_code, request.text))

	def getTags(self, variables):
		listOfNames = []
		query = '''
			    query($owner: String!, $name: String!, $cursor: String, $count: Int!) {
                    repository(owner: $owner, name: $name) {
                        isRepoPrivate: isPrivate
                        payload:
                            refs(
                                first: $count
                                after: $cursor
                                orderBy: {field: TAG_COMMIT_DATE, direction: DESC}
                                refPrefix: \"refs/tags/\"
                            ) {
                                pageInfo {
                                    hasNextPage
                                    endCursor
                                }
                                nodes {
                                    version: name
                                    target {
                                        type: __typename
                                        ... on Commit {
                                        oid
                                        releaseTimestamp: committedDate
                                        }
                                        ... on Tag {
                                            target {
                                                ... on Commit {
                                                oid
                                                }
                                            }
                                            tagger {
                                                releaseTimestamp: date
                                            }
                                        }
                                    }
                                }
                            }
                    }
                }
        '''

		result = self.run_query(query, variables) #execute query
		print(json.dumps(result))

def get_jwt(pem, app_id):
    # Open PEM
    with open(pem, 'rb') as pem_file:
        signing_key = jwt.jwk_from_pem(pem_file.read())

    payload = {
        # Issued at time
        'iat': int(time.time()),
        # JWT expiration time (10 minutes maximum)
        'exp': int(time.time()) + 600,
        # GitHub App's identifier
        'iss': app_id
    }

    # Create JWT
    jwt_instance = jwt.JWT()
    encoded_jwt = jwt_instance.encode(payload, signing_key, alg='RS256')

    return encoded_jwt

########################################
# Set config in the following block
########################################

pem = "key.pem" # Path to the private key of the app
app_id = 123456 # ID of your app
installation_id = 123456 # ID of your app installation

variables = {
    "owner": "bridgecrewio",
    "name": "checkov-action",
    "count": 5,
}
########################################

jwt = get_jwt(pem, app_id)

request = requests.post(f"https://api.github.com/app/installations/{installation_id}/access_tokens", headers={
    "Accept": "application/vnd.github+json",
    "Authorization": f"Bearer {jwt}",
    "X-GitHub-Api-Version": "2022-11-28"
})

token = ""
if request.status_code == 201:
	token = request.json()["token"]
else:
    raise Exception("Query failed to run by returning code of {}. Message {}".format(request.status_code, request.text))

query = graphQL(token)
query.getTags(variables)
```

</details>

**Reply by aalvarezaph** [reply]
@morremeyer were you able to get a solution from Github?

**Reply by morremeyer** [reply]
Still talking to them, the issue is known to the responsible engineering team. No solution or ETA for one yet.

**Reply by morremeyer** [reply]
Got an update by GitHub that basically says "it's in the backlog, but no promises"

> Our engineering team has indicated that this work will not be completed in the near-term.
 >
> With that in mind I'm going to update the status of this support request as Solved. However, please note that the status of Closed or Solved is simply a state within our support system, and does not necessarily reflect the state of the internal issue with our engineering team.

**Comment by morremeyer** [comment]
While I still have an open issue with the GitHub support (see https://github.com/renovatebot/renovate/discussions/17826#discussioncomment-7865235), I've been annoyed enough to look for a workaround.

I found one: Using a host rule that uses a PAT (since these are not affected by this bug)

```
{
  "hostRules": [
    {
      "matchHost": "https://api.github.com/repos/bridgecrewio",
      "token": "FINE_GRAINED_PAT_WITH_PUBLIC_READ_ONLY"
    }
  ]
}
```

**Comment by rarkins** [comment]
Does anyone know if GitHub resolved this problem on their own? This repo seems to work and we haven't turned on fixed IPs in the app yet: https://github.com/nabeelsaabnaTests/pdmp-static-ip-test/pull/3

**Reply by morremeyer** [reply]
They did not resolve the problem. It works for the repo you mentioned because that's a personal repo. The combination that is broken is "org repo" + "GitHub app"

**Reply by rarkins** [reply]
Thanks! So any org repo fails, any personal repo succeeds?

**Reply by morremeyer** [reply]
Yup, that’s the current state of my understanding. If we find any where this differs, it might help the GitHub engineers to debug this issues, but it also seems to be very low on their priority list.

**Comment by flokoe** [comment]
As of today, it seems that the problem is still not solved.

We encountered odd behavior of Renovate opening and auto-closing updates for `bridgecrewio/checkov-action` with the following error message:

```plain
DEBUG: Datasource unknown error
{
  "datasource": "github-tags"
  "packageName": "bridgecrewio/checkov-action"
  "err": {
    "message": "Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and your IP address is not permitted to access this resource.",
    "stack": "Error: Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and your IP address is not permitted to access this resource.\n    at GithubGraphqlDatasourceFetcher.doRawQuery (/usr/local/renovate/lib/util/github/graphql/datasource-fetcher.ts:135:21)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at GithubGraphqlDatasourceFetcher.doShrinkableQuery (/usr/local/renovate/lib/util/github/graphql/datasource-fetcher.ts:200:20)\n    at GithubGraphqlDatasourceFetcher.doPaginatedFetch (/usr/local/renovate/lib/util/github/graphql/datasource-fetcher.ts:244:27)\n    at GithubGraphqlDatasourceFetcher.doCachedQuery (/usr/local/renovate/lib/util/github/graphql/datasource-fetcher.ts:284:7)\n    at GithubGraphqlDatasourceFetcher.getItems (/usr/local/renovate/lib/util/github/graphql/datasource-fetcher.ts:322:17)\n    at Function.query (/usr/local/renovate/lib/util/github/graphql/datasource-fetcher.ts:59:19)\n    at queryTags (/usr/local/renovate/lib/util/github/graphql/index.ts:15:15)\n    at GithubTagsDatasource.getReleases (/usr/local/renovate/lib/modules/datasource/github-tags/index.ts:78:24)\n    at getRegistryReleases (/usr/local/renovate/lib/modules/datasource/index.ts:77:15)\n    at huntRegistries (/usr/local/renovate/lib/modules/datasource/index.ts:118:13)\n    at fetchReleases (/usr/local/renovate/lib/modules/datasource/index.ts:291:15)\n    at lookupUpdates (/usr/local/renovate/lib/workers/repository/process/lookup/index.ts:134:56)\n    at Function.wrap (/usr/local/renovate/lib/util/stats.ts:38:20)\n    at fetchDepUpdates (/usr/local/renovate/lib/workers/repository/process/fetch.ts:58:42)\n    at /usr/local/renovate/lib/workers/repository/process/fetch.ts:104:23\n    at /usr/local/renovate/node_modules/.pnpm/p-map@4.0.0/node_modules/p-map/index.js:57:22"
  }
}
```

**Reply by morremeyer** [reply]
Yup, I'm still poking the GitHub support every now and then, this is just very low on their priority list.

**Reply by morremeyer** [reply]
Update from GitHub in https://github.com/renovatebot/renovate/discussions/17826#discussioncomment-12003779

**Comment by esetnik** [comment]
I am having the same issue in https://github.com/AikidoSec/firewall-php/issues/411

**Comment by esetnik** [comment]
## Root cause & solution

We ran into this same issue with a different organization that has IP allowlisting enabled on their GitHub org.

**Root cause:** The Mend-hosted Renovate app uses the GitHub GraphQL API for `github-releases` and `github-tags` datasource lookups. When an organization enables GitHub's IP allowlist feature, the GraphQL API rejects requests from IPs not on the allowlist — even for public repositories. This is a known GitHub bug where org-level IP restrictions unintentionally affect public repo API access via GraphQL, while the REST API remains unaffected.

**Why the commonly suggested PAT workaround didn't work initially:** Adding a host rule with a PAT via the Mend portal without specifying a `hostType` (or with the wrong one like `git-releases`) doesn't override Mend's built-in `github-releases` and `github-tags` host rules, because Renovate's host rule matching gives precedence to more specific rules (those with a `hostType` set).

**What worked:** In the Mend portal, we created host rules with `hostType` set to **`github-releases`** (the exact datasource name, not `git-releases`) pointing to our PAT. This overrides Mend's platform-injected token for that specific datasource, and since PATs authenticate as a user rather than a GitHub App, they bypass the org-level IP allowlist restriction.

**TL;DR:** If you're hitting this on the Mend-hosted app, add a host rule in the Mend portal with `hostType=github-releases` (and/or `github-tags`), `matchHost=api.github.com`, and a fine-grained PAT with default public read-only access. The host type must exactly match the datasource name — `git-releases` won't work.

**Reply by viceice** [reply]
you should also set the full repo/ org https path as match host, so the pat isn't used for all requests

**Reply by esetnik** [reply]
this can happen on any public repo that exhibits the issue so I guess PAT on all requests is the best solution overall.

**Comment by jamietanna** [comment]
More discussion in https://github.com/renovatebot/renovate/discussions/43238 (but it's more likely to be blocked by GitHub needing to make platform changes)
